cairo-libs/packages/ics23/src/verify.cairo

```diff
@@ -70,8 +69,24 @@ pub fn verify_existence(
 }
 
 pub fn verify_non_existence(
-    spec: @ProofSpec, proof: @NonExistenceProof, root: @RootBytes, key: @KeyBytes,
-) {}
+    spec: @ProofSpec, proof: @NonExistenceProof, root: @RootBytes, key: KeyBytes,
+) {
+    if let Option::Some(left) = proof.left {
+        verify_existence(spec, left, root, left.key, left.value);
+        assert(
+            spec.key_for_comparison(key) > spec.key_for_comparison(left.key.clone()),
+            ICS23Errors::INVALID_LEFT_KEY_ORDER,
+        )
+    } else if let Option::Some(right) = proof.right {
+        verify_existence(spec, right, root, right.key, right.value);
+        assert(
+            spec.key_for_comparison(key) < spec.key_for_comparison(right.key.clone()),
+            ICS23Errors::INVALID_RIGHT_KEY_ORDER,
+        )
+    } else {
```

Comment:
are you sure this is correct ?

you made sure

```
spec.key_for_comparison(left.key.clone()) < spec.key_for_comparison(key) < spec.key_for_comparison(right.key.clone())
```

But you also have to make sure there is nothing between `left` and `right` nodes. That is, `left`'s immediate-right sibling is `right`, and `right`'s immediate-left sibling is `left`.

Reply:
essentially this check in ics23's rust impl: https://github.com/cosmos/ics23/blob/a324422529b8c00ead00b4dcee825867c494cddd/rust/src/verify.rs#L69-L76

Reply:
[c22a4f2](https://github.com/informalsystems/ibc-starknet/pull/305/commits/c22a4f2b464683322b23c58ddeebc3f93f9ff36a)